### ISSUE TITLE
Could you provide the FLUX.2 VAE in Diffusers format?

### ISSUE THREAD

**Issue by CS123n** [madebyollin/taesd#35]
Could you provide the FLUX.2 VAE in Diffusers format?
Thanks a lot!

**Comment by LostRuins** [comment]
seconded, or it would be helpful if you had a conversion script we could use

**Comment by madebyollin** [comment]
# TL;DR and workarounds

TAEF2 has a different architecture than TAEF1 and isn't directly compatible with diffusers' AutoencoderTiny.

1. Short-term, you can use a small bit of wrapper code to make TAEF2 behave like an `AutoencoderTiny` and integrate it into Diffusers pipelines that way. Here's an [example notebook](https://colab.research.google.com/drive/1Cb9ErK-oPOqK6vtE8PvqKYLShbwHvtPd?usp=sharing) using TAEF2 in the FLUX.2 Klein pipeline.
```python
from taesd import TAESD
import torch
from diffusers.utils.accelerate_utils import apply_forward_hook

class DotDict(dict):
    __getattr__ = dict.__getitem__
    __setattr__ = dict.__setitem__

class DiffusersVAEWrapper(torch.nn.Module):
    def __init__(self, taesd):
        super().__init__()
        self.taesd = taesd
        self.dtype = next(self.taesd.parameters()).dtype
        self.bn = torch.nn.BatchNorm2d(128, affine=False, eps=0.0) # default bn
        self.config = DotDict(batch_norm_eps=self.bn.eps)

    @apply_forward_hook
    def encode(self, x):
        return self.taesd.encoder(x.to(self.dtype).mul(0.5).add_(0.5)).to(x.dtype)
    @apply_forward_hook
    def decode(self, x, return_dict=True):
        x = self.taesd.decoder(x.to(self.dtype)).mul(2).sub_(1).clamp_(-1, 1).to(x.dtype)
        return dict(sample=x) if return_dict else x,

taef2_diffusers = DiffusersVAEWrapper(TAESD(encoder_path="taef2_encoder.pth", decoder_path="taef2_decoder.pth").to(torch.bfloat16)).eval().requires_grad_(False)
```

2. Longer-term, we can potentially make a PR for official TAEF2 support in Diffusers' AutoencoderTiny class, and a dedicated HF repo for the converted weights. There might be a way to do this already by specifying a custom model class or something, not sure.

# Longer explanation of why the TAEF2 architecture is different

I trained an initial version of TAEF2 using the TAEF1 architecture, but observed that colors/saturation during training (at 256px) were iffy and colors during inference (at higher resolutions) were way off. The evidence suggested that:
1. FLUX.2 was using its global pooling/norm layers to transmit an overall contrast level per-image, in a way that generalized across resolutions.
2. TAEF2 (initial version) was "baking in" the expected contrast level for 256x256, in a way that didn't generalize that well across images and definitely failed to generalize to higher resolutions.

This is an issue I remembered from prior TAEs (see [reports](https://www.reddit.com/r/StableDiffusion/comments/1p8462z/comment/nr2mwmq/) of "tiny VAE decoding is more saturated than official VAE") but when distilling FLUX.2 VAE the mismatch got much worse.

To address the issue, I added some global-norm-and-pooling layers to the lowest-resolution blocks for TAEF2 specifically (see the [diff](https://github.com/madebyollin/taesd/commit/5cccb12dc72074a4be35680c3b108548a7c69978)). This fixed the saturation/contrast issues (these figures are 1024x1024, with the original architecture/checkpoint marked "No Pool"). However, it broke TAEF2 Diffusers compatibility since Diffusers AutoencoderTiny doesn't include the global norm/pooling layers.

<img width="4160" height="1200" alt="Image" src="https://github.com/user-attachments/assets/b96a3805-4f3c-4ef9-97fd-1f02fd967bb9" />

<img width="4160" height="1200" alt="Image" src="https://github.com/user-attachments/assets/1ef53c52-bb18-4a32-baf9-a4e27b6b1c1c" />

(For posterity, the no-pool weights are in [taef2_no_pool.pth.zip](https://github.com/user-attachments/files/24697832/taef2_no_pool.pth.zip))

**Comment by madebyollin** [comment]
**Update:** For convenience, in lieu of full Diffusers support, I've uploaded a `taef2.safetensors` (using Diffusers' AutoencoderTiny checkpoint structure) + the example wrapper code for using it to https://huggingface.co/madebyollin/taef2/

**Comment by LostRuins** [comment]
works well with https://github.com/leejet/stable-diffusion.cpp/pull/1211

note the different shape in the first layer decoder.layers.0.weight'  [3, 3, 32, 64] vs [3, 3, 16, 64]

I also want to say thank you for not making this as big as taehv. it's a much better size.

**Comment by madebyollin** [comment]
Glad to hear that @LostRuins! The first-layer shape changed because FLUX.2 doubled the number of latent channels to 32. I'll try to keep the model size stable for all future 8x-downscaling image VAEs. Higher-spatial compression VAEs will likely have to be larger (like [TAESANA](https://github.com/madebyollin/taesd/issues/29#issuecomment-3693708271)) in order to understand the increased number of latent channels. I could maybe try using group convs to keep the size down, but haven't tested that idea yet.

One other minor update on the TAEF2 front - I've added some comparison images showing the expected quality vs. other FLUX.2 VAEs here https://huggingface.co/madebyollin/taef2#quality-comparisons 👍

**Comment by rathoreyuvraj** [comment]
> # TL;DR and workarounds
> 
> TAEF2 has a different architecture than TAEF1 and isn't directly compatible with diffusers' AutoencoderTiny.
> 
>     1. Short-term, you can use a small bit of wrapper code to make TAEF2 behave like an `AutoencoderTiny` and integrate it into Diffusers pipelines that way. Here's an [example notebook](https://colab.research.google.com/drive/1Cb9ErK-oPOqK6vtE8PvqKYLShbwHvtPd?usp=sharing) using TAEF2 in the FLUX.2 Klein pipeline.
> 
> 
> from taesd import TAESD
> import torch
> from diffusers.utils.accelerate_utils import apply_forward_hook
> 
> class DotDict(dict):
>     __getattr__ = dict.__getitem__
>     __setattr__ = dict.__setitem__
> 
> class DiffusersVAEWrapper(torch.nn.Module):
>     def __init__(self, taesd):
>         super().__init__()
>         self.taesd = taesd
>         self.dtype = next(self.taesd.parameters()).dtype
>         self.bn = torch.nn.BatchNorm2d(128, affine=False, eps=0.0) # default bn
>         self.config = DotDict(batch_norm_eps=self.bn.eps)
> 
>     @apply_forward_hook
>     def encode(self, x):
>         return self.taesd.encoder(x.to(self.dtype).mul(0.5).add_(0.5)).to(x.dtype)
>     @apply_forward_hook
>     def decode(self, x, return_dict=True):
>         x = self.taesd.decoder(x.to(self.dtype)).mul(2).sub_(1).clamp_(-1, 1).to(x.dtype)
>         return dict(sample=x) if return_dict else x,
> 
> taef2_diffusers = DiffusersVAEWrapper(TAESD(encoder_path="taef2_encoder.pth", decoder_path="taef2_decoder.pth").to(torch.bfloat16)).eval().requires_grad_(False)
> 
>     2. Longer-term, we can potentially make a PR for official TAEF2 support in Diffusers' AutoencoderTiny class, and a dedicated HF repo for the converted weights. There might be a way to do this already by specifying a custom model class or something, not sure.
> 
> 
> # Longer explanation of why the TAEF2 architecture is different
> 
> I trained an initial version of TAEF2 using the TAEF1 architecture, but observed that colors/saturation during training (at 256px) were iffy and colors during inference (at higher resolutions) were way off. The evidence suggested that:
> 
>     1. FLUX.2 was using its global pooling/norm layers to transmit an overall contrast level per-image, in a way that generalized across resolutions.
> 
>     2. TAEF2 (initial version) was "baking in" the expected contrast level for 256x256, in a way that didn't generalize that well across images and definitely failed to generalize to higher resolutions.
> 
> 
> This is an issue I remembered from prior TAEs (see [reports](https://www.reddit.com/r/StableDiffusion/comments/1p8462z/comment/nr2mwmq/) of "tiny VAE decoding is more saturated than official VAE") but when distilling FLUX.2 VAE the mismatch got much worse.
> 
> To address the issue, I added some global-norm-and-pooling layers to the lowest-resolution blocks for TAEF2 specifically (see the [diff](https://github.com/madebyollin/taesd/commit/5cccb12dc72074a4be35680c3b108548a7c69978)). This fixed the saturation/contrast issues (these figures are 1024x1024, with the original architecture/checkpoint marked "No Pool"). However, it broke TAEF2 Diffusers compatibility since Diffusers AutoencoderTiny doesn't include the global norm/pooling layers.
> <img alt="Image" width="2000" height="1200" src="https://private-user-images.githubusercontent.com/8733615/537309353-b96a3805-4f3c-4ef9-97fd-1f02fd967bb9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyOTQ5NTAsIm5iZiI6MTc3MDI5NDY1MCwicGF0aCI6Ii84NzMzNjE1LzUzNzMwOTM1My1iOTZhMzgwNS00ZjNjLTRlZjktOTdmZC0xZjAyZmQ5NjdiYjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMDVUMTIzMDUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjM2Y2NlYzhjZTVmMGQ5MWYyYjY0Yjg3M2UyNGQ0MmM2ZDk0ZjQ5YTVlMGY1MjE2YmE4ODg2OWQwYmM2MzA1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.i1dk4K-ZicgR51wYXS3NbyYFojuC4xTuNpFBdlkmOGc"> <img alt="Image" width="2000" height="1200" src="https://private-user-images.githubusercontent.com/8733615/537309355-1ef53c52-bb18-4a32-baf9-a4e27b6b1c1c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyOTQ5NTAsIm5iZiI6MTc3MDI5NDY1MCwicGF0aCI6Ii84NzMzNjE1LzUzNzMwOTM1NS0xZWY1M2M1Mi1iYjE4LTRhMzItYmFmOS1hNGUyN2I2YjFjMWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMDVUMTIzMDUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzMzZmFjMWU1MDAzMWI2NjIzZjY0ZTc2OGE0NDNiNjVlMTlhZjgwNjk2MjQ2ZWMwZjlmNjcwZTU4MTQyMzQ2YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-pdn1cH4gycK9I76QvoJxSw9kzLL0lxZIqQIQrwl_lY">
> 
> (For posterity, the no-pool weights are in [taef2_no_pool.pth.zip](https://github.com/user-attachments/files/24697832/taef2_no_pool.pth.zip))

Hey can you provide code to compare Flux.2 Encoder with TAEF2 and TAEF2(no pool) as i wanted to compare it for some set of images and the image contrast is off for the outputs.

I am using this kind of code 
```
        latent = self.vae_flux2.encode(x).latent_dist.sample()        

        # decode latent to image
        img_reconstruction = self.vae_flux_2_tiny.decode(latent)

        # save reconstructed image
        save_image(img_reconstruction['sample'] * 0.5 + 0.5, "tmp.png")
```

**Comment by madebyollin** [comment]
@rathoreyuvraj Sure, [here's a notebook](https://colab.research.google.com/drive/1InLMF6PVlXaYvPvcH_rGT7QkUacHjN4b?usp=sharing) to reproduce the comparisons.

<img width="1389" height="534" alt="Image" src="https://github.com/user-attachments/assets/a0e50cde-02aa-4067-a5a7-014a803e324e" />

From the sample code snippet you posted, you're likely missing the latent scale/shift steps that the huggingface VAEs [place in the pipeline code for some reason](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux2/pipeline_flux2.py#L613). BFL's official implementation [performs normalization](https://github.com/black-forest-labs/flux2/blob/main/src/flux2/autoencoder.py#L324) in the `encode`/`decode` methods where it's a lot harder to miss.
